File: src/main/java/com/lostintheway/shareapp/transactions/entity/Transaction.java
package com.lostintheway.shareapp.transactions.entity;

import com.lostintheway.shareapp.portfolios.entity.Portfolio;

import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;

@Entity
public class Transaction {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int id;
    private int shareName;
    private int date;
    private int quantity;
    private int price;
    private int transType;
    private int userId; // no relation

    @ManyToOne
    @JoinColumn(referencedColumnName = "id")
    private Portfolio portfolioId;

    public Transaction(int id, int shareName, int date, int quantity, int price, int transType, int userId) {
        this.id = id;
        this.shareName = shareName;
        this.date = date;
        this.quantity = quantity;
        this.price = price;
        this.transType = transType;
        this.userId = userId;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public int getShareName() {
        return shareName;
    }

    public void setShareName(int shareName) {
        this.shareName = shareName;
    }

    public int getDate() {
        return date;
    }

    public void setDate(int date) {
        this.date = date;
    }

    public int getQuantity() {
        return quantity;
    }

    public void setQuantity(int quantity) {
        this.quantity = quantity;
    }

    public int getPrice() {
        return price;
    }

    public void setPrice(int price) {
        this.price = price;
    }

    public int getTransType() {
        return transType;
    }

    public void setTransType(int transType) {
        this.transType = transType;
    }

    public int getUserId() {
        return userId;
    }

    public void setUserId(int userId) {
        this.userId = userId;
    }

}

File: controllers/transaction.go
package controllers

import (
	"fmt"
	"log"
	"time"

	"github.com/gofiber/fiber/v2"
	"github.com/lostintheway/shareapp/tree/shareapp_go_ent/entconfig"
)

func POSTtransaction(ctx *fiber.Ctx) error {

	body := new(struct {
		ShareName string `json:"shareName"`
		Date      string `json:"date"`
		Quantity  int    `json:"quantity"`
		Price     int    `json:"price"`
		TransType string `json:"transType"`
		UserId    int    `json:"userId"`
		PfId      int    `json:"pfId"`
	})

	if err := ctx.BodyParser(&body); err != nil {
		return err
	}

	mydate, err := time.Parse("2006-01-02", body.Date)
	if err != nil {
		log.Fatal(err)
	}
	transaction, err := entconfig.Client.Transaction.
		Create().
		SetShareName(body.ShareName).
		SetDate(mydate).
		SetQuantity((body.Quantity)).
		SetPrice((body.Price)).
		SetTransType((body.TransType)).
		SetUserID((body.UserId)).
		SetPfIDID((body.PfId)).
		Save(ctx.Context())
	if err != nil {
		return fmt.Errorf("error post transaction: %w", err)
	}
	log.Println("transactions created: ", transaction)
	return ctx.Status(200).JSON(transaction)
}

func GETAllTransactions(ctx *fiber.Ctx) error {
	transactions, err := entconfig.Client.Transaction.Query().All(ctx.Context())
	if err != nil {
		return fiber.NewError(500, err.Error())
	}
	log.Println("returned battles:", transactions)

	return ctx.Status(200).JSON(transactions)
}
